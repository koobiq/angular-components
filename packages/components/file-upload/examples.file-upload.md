Пример демонстрирует возможности локализации компонента по месту

<!-- example(file-upload-multiple-custom-text-overview) -->

Пример загрузки файлов с неопределенным по завершенности прогрессом:

<!-- example(file-upload-indeterminate-loading-overview) -->

Пример использования с Control Value Accessor:

<!-- example(file-upload-cva-overview) -->
