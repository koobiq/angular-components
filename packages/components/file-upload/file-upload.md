Позволяет пользователю загрузить файлы в продукт.

<!-- example(file-upload-single-overview) -->

### Когда использовать

Используйте, когда необходимо загрузить один или несколько файлов.

Поле загрузки файлов можно использовать в форме или вне формы как самостоятельный элемент.

### Структура компонента

Загрузка файлов состоит из:

##### Когда файлы не выбраны

* Зоны, в которую можно перетащить файл(ы) для загрузки
* Ссылки, с помощью которой можно выбрать файл(ы) для загрузки
* Текста подсказки (не обязательно)

<div class="kbq-alert kbq-alert_info" style="margin-top: 15px;">
    <i class="mc kbq-icon kbq-info-o_16 kbq-alert__icon"></i>
    <span> Если есть ограничения по формату допустимых файлов или по размеру, перечислите их в тексте подсказки </span>
</div>

##### Файлы загружены

* Иконки [файла](https://koobiq.io/icons?id=file-empty_16)
* Имени загруженного файла
* Размер загруженного файла (только при загрузки нескольких файлов)
* Иконки [удалить](https://koobiq.io/icons?id=close-circle_16)

#### Вариации

##### Загрузка одного файла

<!-- example(file-upload-single-overview) -->

##### Загрузка нескольких файлов

Размер по умолчанию

<!-- example(file-upload-multiple-default-overview) -->

Компактный размер (внешне то же самое, что и загрузка одного файла, отличие только в тексте)

<!-- example(file-upload-multiple-compact-overview) -->

### Описание работы

Файл(ы) можно перетаскивать в обозначенную зону пунктирной линией. Или выбирать через проводник операционной системы на устройстве.

##### Загрузка одного файла

После того как файл выбран, компонент получает заблокированное состояние с индикатором загрузки [спиннер](https://koobiq.io/components/progress-spinner/overview), до того момента, пока загрузка файла не будет завершена, но при этом остается возможность отменить загрузку путем нажатия на иконку очистить.

Когда файл загружен, компонент получает нормальное состояние, для приложенного файла подставляется иконка, соответствующая типу файла (делается на стороне продукта, техническая возможность в компонент заложена) и подставляется имя файла.

Чтобы приложить другой файл, его можно перетащить в обозначенную зону пунктирной линией.

Если названия файла не влезает полностью, оно в середине обрезается многоточием, а при наведении показывается [тултип](https://koobiq.io/components/tooltip/overview) с полным названием (это правило действует и для загрузки нескольких файлов)

<div class="kbq-alert kbq-alert_warning" style="margin-top: 15px;">
    <i class="mc kbq-icon kbq-error_16 kbq-alert__icon"></i>
    <span> Стандартный тултип из Koobiq имеет ограничение по ширине в 240px, в данном случае стоит это ограничение не учитывать, чтобы имя файла показывалось в тултипе без переносов. Но ширина тултипа не должна превышать ширину экрана, в таком случае перенос имени файла на другую строку допустим. </span>
</div>

##### Загрузка нескольких файлов

После того как файлы выбраны, рядом с ними появляются индикаторы загрузки [спиннеры]((https://koobiq.io/components/progress-spinner/overview)), до того момента, пока загрузка файлов не будет завершена.

Когда файлы загружены, рядом с приложенными файлами подставляются иконки, соответствующие типу файлов (делается на стороне продукта, техническая возможность в компонент заложена) и подставляются имена файлов.

Возможность перетаскивать и выбирать файлы остается.

#### Валидация

<div class="kbq-alert kbq-alert_warning" style="margin-top: 15px;">
    <i class="mc kbq-icon kbq-error_16 kbq-alert__icon"></i>
    <span> Если процесс загрузки файлов не завершился, ограничьте возможность отправки формы путем блокирования кнопки отправки. </span>
</div>

##### Загрузка одного файла

Текст ошибки выводится под компонентом, с указанием причины ошибки.

Если не удалось загрузить файл по неизвестным причинам, вместо иконки [файла](https://koobiq.io/icons?id=file-empty_16), выводится иконка [ошибки](https://koobiq.design/icons?id=error_16)

<!-- example(file-upload-single-error-overview) -->

##### Загрузка нескольких файлов

Текст ошибки выводится под компонентом.

Причины ошибок перечисляются для каждого файла отдельно.

<!-- example(file-upload-multiple-error-overview) -->

#### Фокус и работа с клавиатурой

##### Загрузка одного файла

| Файл не выбран                                                                      |                                                                                                      |
|-------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|
| <span class="hot-key-button">Space</span> \ <span class="hot-key-button">↵</span>   | Если фокус установлен на ссылке загрузке файлов → Открывает диалог выбора файла операционной системы |
| <span class="hot-key-button">Tab</span>                                             | Если фокус установлен на ссылке загрузке файлов → Переход к следующему компоненту в tab sequence     |
| <span class="hot-key-button">Shift</span> + <span class="hot-key-button">Tab</span> | Если фокус установлен на ссылке загрузке файлов → Переход к предыдущему компоненту в tab sequence    |

| Файл выбран                                                                                                                    |                                                                                                   |
|--------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------|
| <span class="hot-key-button">Space</span> \ <span class="hot-key-button">↵</span> \ <span class="hot-key-button">Delete</span> | Если фокус установлен на иконке удаления файлов → Удаляет выбранный файл                          |
| <span class="hot-key-button">Tab</span>                                                                                        | Если фокус установлен на иконке удаления файлов → Переход к следующему компоненту в tab sequence  |
| <span class="hot-key-button">Shift</span> + <span class="hot-key-button">Tab</span>                                            | Если фокус установлен на иконке удаления файлов → Переход к предыдущему компоненту в tab sequence |

##### Загрузка нескольких файлов

| Файлы не выбраны                                                                  |                                                                                                      |
|-----------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|
| <span class="hot-key-button">Space</span> \ <span class="hot-key-button">↵</span> | Если фокус установлен на ссылке загрузке файлов → Открывает диалог выбора файла операционной системы |

| Файлы выбраны                                                                       |                                                                                                                                                                      |
|-------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <span class="hot-key-button">Space</span> \ <span class="hot-key-button">↵</span>   | Если фокус установлен на ссылке загрузке файлов → Открывает диалог выбора файла операционной системы                                                                 |
| <span class="hot-key-button">Delete</span> \                                        | Если фокус установлен на списке файлов → Удаляет выделенный файл в списке                                                                                            |
| <span class="hot-key-button">Tab</span>                                             | Если фокус установлен на списке файлов → Переход к загрузке файлов. Если фокус установлен на ссылке загрузке файлов → Переход к следующему компоненту в tab sequence |
| <span class="hot-key-button">Shift</span> + <span class="hot-key-button">Tab</span> | Если фокус установлен на списке файлов → Переход к предыдущему компоненту в tab sequence. Если фокус установлен на ссылке загрузке файлов → Переход к списку файлов  |
| <span class="hot-key-button">↑</span> <span class="hot-key-button">↓</span>         | Если фокус установлен на списке файлов → Перемещение между элементами списка выбранных файлов                                                                        |

### Дизайн

#### Размер

##### Загрузка одного файла

Ширину выбирает дизайнер.
Высота фиксированная.

##### Загрузка нескольких файлов

**Файлы не выбраны**

Ширину и высоту выбирает дизайнер.

**Файлы выбраны**

Ширину выбирает дизайнер.
Высота соответствует выбранной, когда файлы не выбраны, после увеличивается в зависимости от количества выбранных файлов.

### Единицы измерения размера файла

Используйте правило из положения о единицах величин, допускаемых к применению в Российской Федерации:

> Наименование и обозначение единицы количества информации "байт" (1 байт = 8 бит)
применяются с двоичными приставками "Кило", "Мега", "Гига", которые соответствуют
множителям "2^10", "2^20" и "2^30" (1 Кбайт = 1024 байт, 1 Мбайт = 1024 Кбайт, 1 Гбайт = 1024
Мбайт).
