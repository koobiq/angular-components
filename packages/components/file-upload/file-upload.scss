@use 'sass:meta';
@use 'sass:map';

.kbq-file-upload {
    box-sizing: border-box;
    display: flex;
    align-items: center;
    position: relative;

    .dropzone,
    .file-item {
        display: flex;
        align-items: center;
    }

    .kbq-ellipsis-center {
        position: relative;
        display: flex;

        .data-text-start {
            flex: 0 1 auto;
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: pre;
        }

        .data-text-end {
            flex: 1 0 auto;
            overflow: hidden;
            white-space: pre;
        }
    }
}
