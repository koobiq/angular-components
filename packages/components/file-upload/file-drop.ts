import { Directive, EventEmitter, Output } from '@angular/core';
import { KbqFile } from './file-upload';

const isFolderCanBeDragged = (): boolean => 'webkitGetAsEntry' in DataTransferItem.prototype;
const entryIsDirectory = (entry?: FileSystemEntry): entry is FileSystemDirectoryEntry => !!entry && entry.isDirectory;
const entryIsFile = (entry?: FileSystemEntry): entry is FileSystemFileEntry => !!entry && entry.isFile;

@Directive({
    selector: '[kbqFileDrop]',
    exportAs: 'kbqFileDrop',
    host: {
        '[class.dragover]': 'dragover',
        '(dragover)': 'onDragOver($event)',
        '(dragleave)': 'onDragLeave($event)',
        '(drop)': 'onDrop($event)'
    }
})
export class KbqFileDropDirective {
    dragover: boolean;

    @Output() filesDropped: EventEmitter<FileList | KbqFile[]> = new EventEmitter<FileList | KbqFile[]>();

    onDragOver(event: DragEvent) {
        event.preventDefault();
        event.stopPropagation();
        this.dragover = true;
    }

    onDragLeave(event: DragEvent) {
        event.preventDefault();
        event.stopPropagation();
        this.dragover = false;
    }

    onDrop(event: DragEvent) {
        if (!isFolderCanBeDragged()) {
            console.warn('Drag-and-drop functionality for folders is not supported by this browser.');
        }

        event.preventDefault();
        event.stopPropagation();
        this.dragover = false;

        if (event.dataTransfer && event.dataTransfer.items.length > 0) {
            // event.dataTransfer.items requires dom.iterable lib
            // @ts-ignore
            const fileEntries: FileSystemEntry[] = [...event.dataTransfer.items]
                .filter((item: DataTransferItem) => item.kind === 'file')
                .map((item) => item.webkitGetAsEntry() as FileSystemEntry);

            Promise.all(fileEntries.map(unwrapDirectory))
                .then((fileList) => fileList.reduce((res, next) => res.concat(next), []))
                .then((entries: KbqFile[]) => this.filesDropped.emit(entries));
        }
    }
}

const unwrapDirectory = async (item: FileSystemEntry): Promise<KbqFile[]> => {
    const queue: (FileSystemEntry | Promise<FileSystemEntry[]>)[] = [item];
    const result: Promise<KbqFile>[] = [];

    while (queue.length > 0) {
        const next = queue.pop();
        if (next instanceof Promise) {
            queue.push(...(await next));
        } else if (entryIsDirectory(next)) {
            const directoryReader = next.createReader();

            queue.push(
                new Promise<FileSystemEntry[]>((resolve, reject) => directoryReader.readEntries(resolve, reject))
            );
        } else if (entryIsFile(next)) {
            const fileEntry = next;
            result.push(
                new Promise((resolve, reject) => {
                    fileEntry.file((file) => {
                        (file as KbqFile).fullPath = fileEntry.fullPath;
                        resolve(file as KbqFile);
                    }, reject);
                })
            );
        }
    }

    return Promise.all(result);
};
