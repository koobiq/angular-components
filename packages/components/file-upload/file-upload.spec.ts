import { Component, ElementRef, ViewChild } from '@angular/core';
import { ComponentFixture, TestBed, fakeAsync, flush } from '@angular/core/testing';
import { FormControl, FormsModule, ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { DELETE, TAB } from '@koobiq/cdk/keycodes';
import {
    createFakeEvent,
    dispatchEvent,
    dispatchFakeEvent,
    dispatchKeyboardEvent,
    dispatchMouseEvent
} from '@koobiq/cdk/testing';
import { createFile } from './file-drop.spec';
import { KbqFileItem, KbqFileValidatorFn } from './file-upload';
import { KbqFileUploadModule } from './file-upload.module';
import { KbqMultipleFileUploadComponent } from './multiple-file-upload.component';
import { KbqSingleFileUploadComponent } from './single-file-upload.component';

export function dispatchDropEvent(
    component: any,
    fixture: ComponentFixture<any>,
    fileName = 'test.file',
    type?: string
) {
    const fakeDropEvent = createFakeEvent('drop');
    const fakeItem = createFile(fileName, type);
    (fakeDropEvent as any).dataTransfer = { items: [fakeItem] };

    dispatchEvent(component.elementRef.nativeElement.querySelector('.kbq-file-upload'), fakeDropEvent);
    fixture.detectChanges();
}

const fileItemActionCssClass = 'kbq-file-upload__action';
const fileItemRowCssClass = 'multiple__uploaded-item';

const fileItemCssClass = 'file-item';
const fileItemTextCssClass = 'file-item__text';

const maxFileExceeded = (file: File): string | null => {
    const kilo = 1024;
    const mega = kilo * kilo;
    const maxMbytes = 5;
    const maxSize = maxMbytes * mega;

    return maxSize !== undefined && (file?.size ?? 0) > maxSize ? `Exceeded with (${maxSize / mega} Mb)` : null;
};

describe('MultipleFileUploadComponent', () => {
    let component: any;
    let fixture: ComponentFixture<any>;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [NoopAnimationsModule, KbqFileUploadModule, FormsModule, ReactiveFormsModule],
            declarations: [BasicMultipleFileUpload, ControlValueAccessorMultipleFileUpload]
        }).compileComponents();

        fixture = TestBed.createComponent(BasicMultipleFileUpload);
        component = fixture.componentInstance;
        fixture.detectChanges();
    });

    describe('with focus and keyboard', () => {
        it('should toggle label focus state on input focused/blurred', fakeAsync(() => {
            const fileInput: HTMLInputElement = component.fileUpload.input.nativeElement;

            // Simulate focus via keyboard.
            dispatchKeyboardEvent(fixture.nativeElement, 'keydown', TAB);
            fileInput.focus();
            fixture.detectChanges();
            flush();

            const label: HTMLLabelElement = fixture.nativeElement.querySelector('label');

            expect(label.classList.contains('cdk-keyboard-focused')).toBeTrue();

            dispatchKeyboardEvent(fixture.nativeElement, 'keydown', TAB);
            fileInput.blur();
            fixture.detectChanges();
            flush();

            expect(label.classList.contains('cdk-keyboard-focused')).toBeFalse();
        }));

        it('should NOT toggle label focus state on input focus if disabled', fakeAsync(() => {
            const fileInput: HTMLInputElement = component.fileUpload.input.nativeElement;
            component.disabled = true;
            fixture.detectChanges();

            // Simulate focus via keyboard.
            dispatchKeyboardEvent(fixture.nativeElement, 'keydown', TAB);
            dispatchFakeEvent(fileInput, 'focusin');
            fixture.detectChanges();
            flush();

            const label: HTMLLabelElement = fixture.nativeElement.querySelector('label');

            expect(label.classList.contains('cdk-keyboard-focused')).toBeFalse();
        }));

        it('should remove file via button keydown.delete in a row', () => {
            spyOn(component, 'onChange');
            component.disabled = false;
            fixture.detectChanges();
            const event = createFakeEvent('change');

            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            const subscription = component.fileUpload.fileQueueChanged.subscribe((value) => {
                expect(value.length).toBeFalsy();
            });

            dispatchKeyboardEvent(
                fixture.debugElement.query(By.css(`.${fileItemRowCssClass}`)).nativeElement,
                'keydown',
                DELETE
            );
            subscription.unsubscribe();
        });
    });

    describe('with file queue change', () => {
        it('should add files via input click', () => {
            spyOn(component, 'onChange');
            component.disabled = false;
            fixture.detectChanges();

            const event = createFakeEvent('change');

            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            expect(component.onChange).toHaveBeenCalled();
            const subscription = component.fileUpload.fileQueueChanged.subscribe((value) => {
                expect(value.length).toBeTruthy();
            });

            subscription.unsubscribe();
        });

        it('should NOT add files via input click if disabled', () => {
            spyOn(component, 'onChange');
            component.disabled = true;
            fixture.detectChanges();
            const subscription = component.fileUpload.fileQueueChanged.subscribe((value) => {
                expect(value.length).toBeFalsy();
            });

            const event = createFakeEvent('change');
            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });
            dispatchEvent(component.fileUpload.input.nativeElement, event);

            expect(component.onChange).not.toHaveBeenCalled();
            subscription.unsubscribe();
        });

        it('should add files via drag-n-drop', fakeAsync(() => {
            spyOn(component, 'onChange');
            component.disabled = false;
            fixture.detectChanges();
            const subscription = component.fileUpload.fileQueueChanged.subscribe((value) => {
                expect(value.length).toBeTruthy();
            });

            dispatchDropEvent(component, fixture);
            fixture.detectChanges();
            flush();

            expect(component.onChange).toHaveBeenCalled();
            subscription.unsubscribe();
        }));

        it('should NOT add files via drag-n-drop if disabled', () => {
            spyOn(component, 'onChange');
            component.disabled = true;
            fixture.detectChanges();
            const subscription = component.fileUpload.fileQueueChanged.subscribe((value) => {
                expect(value.length).toBeFalsy();
            });

            dispatchDropEvent(component, fixture);

            expect(component.onChange).not.toHaveBeenCalled();
            subscription.unsubscribe();
        });

        it('should remove file via button click in a row', () => {
            spyOn(component, 'onChange');
            component.disabled = false;
            fixture.detectChanges();

            const event = createFakeEvent('change');

            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            const subscription = component.fileUpload.fileQueueChanged.subscribe((value) => {
                expect(value.length).toBeFalsy();
            });

            fixture.debugElement.query(By.css(`.${fileItemActionCssClass} .kbq-icon`)).nativeElement.click();
            fixture.detectChanges();
            subscription.unsubscribe();
        });
    });

    describe('with custom validation', () => {
        it('should mark added file with error', () => {
            expect(component.files).toBeFalsy();
            // max file === 5e6
            component.validation = [maxFileExceeded];
            fixture.detectChanges();

            const event = createFakeEvent('change');

            const fakeFile: Partial<File> = { name: 'test.file', type: 'test', size: 6e6 };
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            expect(component.files.length).toBeTruthy();
            expect(component.files[0].hasError).toBeTruthy();
        });
    });

    describe('with ControlValueAccessor', () => {
        beforeEach(() => {
            fixture = TestBed.createComponent(ControlValueAccessorMultipleFileUpload);
            component = fixture.componentInstance;
            fixture.detectChanges();
        });

        it('should toggle the disabled state', () => {
            expect(component.fileUpload.disabled).toBe(false);

            component.control.disable();
            fixture.detectChanges();

            expect(component.fileUpload.disabled).toBe(true);

            component.control.enable();
            fixture.detectChanges();

            expect(component.fileUpload.disabled).toBe(false);
        });

        it('should update file value with setValue', () => {
            expect(component.fileUpload.files.length).toBeFalsy();

            const fakeFile: Partial<File> = new File(['test'], 'test');
            const dt = new DataTransfer();
            dt.items.add(fakeFile as File);

            component.control.setValue(dt.files);

            expect(component.fileUpload.files.length).toBeTruthy();
            expect(component.files.length).toBeTruthy();
        });

        it('should update form control touched on file dropped', fakeAsync(() => {
            expect(component.control.touched).toBeFalse();

            dispatchDropEvent(component, fixture);
            fixture.detectChanges();
            flush();

            expect(component.control.touched).toBeTrue();
        }));

        it('should update form control touched on file added via click', () => {
            expect(component.control.touched).toBeFalse();

            const event = createFakeEvent('change');

            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });
            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            expect(component.control.touched).toBeTrue();
        });
    });
});

describe('SingleFileUploadComponent', () => {
    let component: any;
    let fixture: ComponentFixture<any>;
    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [NoopAnimationsModule, KbqFileUploadModule, FormsModule, ReactiveFormsModule],
            declarations: [
                BasicSingleFileUpload,
                ControlValueAccessorSingleFileUpload
            ]
        }).compileComponents();

        fixture = TestBed.createComponent(BasicSingleFileUpload);
        component = fixture.componentInstance;
        fixture.detectChanges();
    });

    describe('with focus and keyboard', () => {
        it('should toggle label focus state on input focused/blurred', fakeAsync(() => {
            const fileInput: HTMLInputElement = component.fileUpload.input.nativeElement;

            // Simulate focus via keyboard.
            dispatchKeyboardEvent(fixture.nativeElement, 'keydown', TAB);
            fileInput.focus();
            fixture.detectChanges();
            flush();

            const label: HTMLLabelElement = fixture.nativeElement.querySelector('label');

            expect(label.classList.contains('cdk-keyboard-focused')).toBeTrue();

            dispatchKeyboardEvent(fixture.nativeElement, 'keydown', TAB);
            fileInput.blur();
            fixture.detectChanges();
            flush();

            expect(label.classList.contains('cdk-keyboard-focused')).toBeFalse();
        }));

        it('should NOT toggle label focus state on input focus if disabled', fakeAsync(() => {
            const fileInput: HTMLInputElement = component.fileUpload.input.nativeElement;
            component.disabled = true;
            fixture.detectChanges();

            // Simulate focus via keyboard.
            dispatchKeyboardEvent(fixture.nativeElement, 'keydown', TAB);
            dispatchFakeEvent(fileInput, 'focusin');
            fixture.detectChanges();
            flush();

            const label: HTMLLabelElement = fixture.nativeElement.querySelector('label');

            expect(label.classList.contains('cdk-keyboard-focused')).toBeFalse();
        }));

        it('should remove file via button keydown.delete', () => {
            component.disabled = false;
            fixture.detectChanges();
            const event = createFakeEvent('change');
            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });
            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            const subscription = component.fileUpload.fileQueueChange.subscribe((value) => {
                expect(value).toBeFalsy();
            });

            dispatchKeyboardEvent(
                component.elementRef.nativeElement.querySelector(`.${fileItemCssClass} .kbq-icon-button`),
                'keydown',
                DELETE
            );
            subscription.unsubscribe();
        });
    });

    describe('with file queue change', () => {
        it('should add file via input click', () => {
            spyOn(component, 'onChange');
            component.disabled = false;
            fixture.detectChanges();

            const event = createFakeEvent('change');

            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            expect(component.onChange).toHaveBeenCalled();
            const subscription = component.fileUpload.fileQueueChange.subscribe((value) => {
                expect(value).toBeTruthy();
            });

            subscription.unsubscribe();
        });

        it('should NOT add file via input click if disabled', () => {
            spyOn(component, 'onChange');
            component.disabled = true;
            fixture.detectChanges();

            const event = createFakeEvent('change');
            Object.defineProperty(event, 'target', { get: () => ({ files: [{ name: 'test' }] }) });
            dispatchEvent(component.fileUpload.input.nativeElement, event);

            expect(component.onChange).not.toHaveBeenCalled();
            expect(component.file).toBeFalsy();
        });

        it('should add file via drag-n-drop', fakeAsync(() => {
            spyOn(component, 'onChange');
            component.disabled = false;
            fixture.detectChanges();

            dispatchDropEvent(component, fixture);
            fixture.detectChanges();
            flush();

            expect(component.onChange).toHaveBeenCalled();
            const subscription = component.fileUpload.fileQueueChange.subscribe((value) => {
                expect(value).toBeTruthy();
            });

            subscription.unsubscribe();
        }));

        it('should NOT add file via drag-n-drop if disabled', () => {
            spyOn(component, 'onChange');
            component.disabled = true;
            fixture.detectChanges();

            dispatchDropEvent(component, fixture);

            expect(component.file).toBeFalsy();
            expect(component.onChange).not.toHaveBeenCalled();
        });

        it('should remove file via button click', () => {
            component.disabled = false;
            fixture.detectChanges();

            const event = createFakeEvent('change');

            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            const subscription = component.fileUpload.fileQueueChange.subscribe((value) => {
                expect(value).toBeFalsy();
            });

            component.elementRef.nativeElement.querySelector(`.${fileItemCssClass} .kbq-icon-button`).click();
            subscription.unsubscribe();
        });
    });

    describe('with ellipsis in the center', () => {
        it('should add tooltip and ellipsis in the center for a file with a long name', fakeAsync(() => {
            component.disabled = false;
            fixture.detectChanges();

            const event = createFakeEvent('change');
            const fakeFile = new File(['test'], 'very very very very very very very very very long file name.txt');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });
            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();
            flush();

            dispatchMouseEvent(
                fixture.debugElement.query(By.css(`.${fileItemTextCssClass}`)).nativeElement,
                'mouseenter'
            );
            fixture.detectChanges();
            flush();

            const tooltipInstance = document.querySelector('.kbq-tooltip');
            expect(tooltipInstance).toBeTruthy();
        }));
    });

    describe('with custom validation', () => {
        it('should mark added file with error', () => {
            expect(component.file).toBeFalsy();
            // max file === 5e6
            component.validation = [maxFileExceeded];
            fixture.detectChanges();

            const event = createFakeEvent('change');

            const fakeFile: Partial<File> = { name: 'test.file', type: 'test', size: 6e6 };
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            expect(component.file).toBeTruthy();
            expect(component.file?.hasError).toBeTruthy();
        });
    });

    describe('with ControlValueAccessor', () => {
        beforeEach(() => {
            fixture = TestBed.createComponent(ControlValueAccessorSingleFileUpload);
            component = fixture.componentInstance;
            fixture.detectChanges();
        });

        it('should toggle the disabled state', () => {
            expect(component.fileUpload.disabled).toBe(false);

            component.control.disable();
            fixture.detectChanges();

            expect(component.fileUpload.disabled).toBe(true);

            component.control.enable();
            fixture.detectChanges();

            expect(component.fileUpload.disabled).toBe(false);
        });

        it('should update file value with setValue', () => {
            expect(component.fileUpload.file).toBeFalsy();

            const fakeFile = new File(['test'], 'test.file');
            component.control.setValue(fakeFile);

            expect(component.fileUpload.file).toBeTruthy();
            expect(component.file).toBeTruthy();
        });

        it('should update form control touched on file dropped', fakeAsync(() => {
            expect(component.control.touched).toBeFalse();

            dispatchDropEvent(component, fixture);
            fixture.detectChanges();
            flush();

            expect(component.control.touched).toBeTrue();
        }));

        it('should update form control touched on file added via click', () => {
            expect(component.control.touched).toBeFalse();

            const event = createFakeEvent('change');

            const fakeFile = new File(['test'], 'test.file');
            Object.defineProperty(event, 'target', { get: () => ({ files: [fakeFile] }) });

            dispatchEvent(component.fileUpload.input.nativeElement, event);
            fixture.detectChanges();

            expect(component.control.touched).toBeTrue();
        });
    });

    // TODO: real-life scenario & test results with the same data are different
    xdescribe('with accepted files list', () => {
        it('should filter files via drag-n-drop with extensions', fakeAsync(() => {
            spyOn(component, 'onChange');
            component.disabled = false;
            component.accept = ['.pdf', '.png'];
            fixture.detectChanges();

            dispatchDropEvent(component, fixture, 'test.test');
            fixture.detectChanges();
            flush();

            dispatchDropEvent(component, fixture, 'test.pdf');
            fixture.detectChanges();
            flush();

            dispatchDropEvent(component, fixture, 'test.png');
            fixture.detectChanges();
            flush();

            expect(component.onChange).toHaveBeenCalledTimes(2);
        }));

        it('should filter files via drag-n-drop with mimeType', fakeAsync(() => {
            spyOn(component, 'onChange');
            component.disabled = false;
            component.accept = ['application/pdf'];
            fixture.detectChanges();

            dispatchDropEvent(component, fixture, 'test.test');
            fixture.detectChanges();
            flush();

            // in file system file type will be automatically provided
            dispatchDropEvent(component, fixture, 'test.pdf', 'application/pdf');
            fixture.detectChanges();
            flush();

            expect(component.onChange).toHaveBeenCalledTimes(1);
        }));
    });
});

@Component({
    selector: '',
    template: `
        <div style="max-width: 350px;">
            <kbq-single-file-upload
                #fileUpload
                [accept]="accept"
                [customValidation]="validation"
                [disabled]="disabled"
                (fileQueueChange)="onChange($event)"
            />
        </div>
    `
})
class BasicSingleFileUpload {
    @ViewChild('fileUpload') fileUpload: KbqSingleFileUploadComponent;
    disabled: boolean;
    file: KbqFileItem | null;
    accept: string[] = [];
    validation: KbqFileValidatorFn[] = [];

    constructor(public elementRef: ElementRef) {}

    onChange(event: KbqFileItem | null) {
        this.file = event;
    }
}

@Component({
    selector: '',
    template: `
        <div style="max-width: 350px;">
            <kbq-file-upload
                #fileUpload
                [formControl]="control"
                [accept]="accept"
                [customValidation]="validation"
                (fileQueueChange)="onChange($event)"
            />
        </div>
    `
})
class ControlValueAccessorSingleFileUpload {
    @ViewChild('fileUpload') fileUpload: KbqSingleFileUploadComponent;
    file: KbqFileItem | null;
    accept: string[] = [];
    validation: KbqFileValidatorFn[] = [];
    control = new FormControl();

    constructor(public elementRef: ElementRef) {}

    onChange(event: KbqFileItem | null) {
        this.file = event;
    }
}

@Component({
    selector: '',
    template: `
        <div style="max-width: 350px;">
            <kbq-multiple-file-upload
                #fileUpload
                [disabled]="disabled"
                [customValidation]="validation"
                (fileQueueChanged)="onChange($event)"
            />
        </div>
    `
})
class BasicMultipleFileUpload {
    @ViewChild('fileUpload') fileUpload: KbqMultipleFileUploadComponent;
    disabled: boolean;
    files: KbqFileItem[];
    validation: KbqFileValidatorFn[] = [];

    constructor(public elementRef: ElementRef) {}

    onChange(event: KbqFileItem[]) {
        this.files = event;
    }
}

@Component({
    selector: '',
    template: `
        <div style="max-width: 350px;">
            <kbq-multiple-file-upload
                #fileUpload
                [formControl]="control"
                [accept]="accept"
                [customValidation]="validation"
                (fileQueueChanged)="onChange($event)"
            />
        </div>
    `
})
class ControlValueAccessorMultipleFileUpload {
    @ViewChild('fileUpload') fileUpload: KbqMultipleFileUploadComponent;
    files: KbqFileItem[];
    accept: string[] = [];
    validation: KbqFileValidatorFn[] = [];
    control = new FormControl();

    constructor(public elementRef: ElementRef) {}

    onChange(event: KbqFileItem[]) {
        this.files = event;
    }
}
