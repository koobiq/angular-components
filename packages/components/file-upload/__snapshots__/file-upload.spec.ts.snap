// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbqMultipleFileUploadComponent with localeConfig input property should use default properties if they not provided with localeConfig 1`] = `
{
  "browseLink": "выберите",
  "captionText": "TEST {{ browseLink }}",
  "captionTextForCompactSize": "Перетащите файлы или {{ browseLink }}",
  "captionTextWhenSelected": "Перетащите еще или {{ browseLink }}",
  "gridHeaders": {
    "file": "Файл",
    "size": "Размер",
  },
  "title": "Загрузите файлы",
}
`;

exports[`KbqSingleFileUploadComponent with localeConfig input property should use default properties if they not provided with localeConfig 1`] = `
{
  "browseLink": "выберите",
  "captionText": "TEST {{ browseLink }}",
}
`;
